server/src/main/java/org/elasticsearch/search/sort/BucketedSort.java

```diff
@@ -0,0 +1,281 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.sort;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Scorable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.lucene.ScorerAware;
+import org.elasticsearch.common.util.BigArray;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.DoubleArray;
+import org.elasticsearch.common.util.FloatArray;
+import org.elasticsearch.common.util.LongArray;
+import org.elasticsearch.search.DocValueFormat;
+
+import java.io.IOException;
+
+/**
+ * Type specialized sort implementations designed for use in aggregations.
+ */
+public abstract class BucketedSort implements Releasable {
+    // TODO priority queue semantics to support mulitiple hits in the buckets
+    protected final BigArrays bigArrays;
+    private final SortOrder order;
+    private final DocValueFormat format;
+    private long maxBucket = -1;
+
+    public BucketedSort(BigArrays bigArrays, SortOrder order, DocValueFormat format) {
+        this.bigArrays = bigArrays;
+        this.order = order;
+        this.format = format;
+    }
+
+    /**
+     * The order of the sort.
+     */
+    public final SortOrder getOrder() {
+        return order;
+    }
+
+    /**
+     * The format to use when presenting the results.
+     */
+    public final DocValueFormat getFormat() {
+        return format;
+    }
+
+    /**
+     * Get the value for a bucket.
+     */
+    public final Object getValue(long bucket) {
+        if (bucket > maxBucket) {
+            return null;
+        }
+        return getValueForBucket(bucket);
+    }
+
+    /**
+     * Get the {@linkplain Leaf} implementation that'll do that actual collecting.
+     */
+    public abstract Leaf forLeaf(LeafReaderContext ctx) throws IOException;
+
+    /**
+     * Does this sort need scores? Most don't, but sorting on {@code _score} does.
+     */
+    public abstract boolean needsScores();
+
+    /**
+     * The {@linkplain BigArray} backing this sort.
+     */
+    protected abstract BigArray buckets();
+
+    /**
+     * Grow the {@linkplain BigArray} backing this sort to account for new buckets.
+     */
+    protected abstract void grow(long minSize);
+
+    /**
+     * Get the value for a bucket. This will only be called if the bucket was hit.
+     */
+    protected abstract Object getValueForBucket(long bucket);
+
+    @Override
+    public void close() {
+        buckets().close();
+    }
+
+    /**
+     * Performs the actual collection against a {@linkplain LeafReaderContext}.
+     */
+    public abstract class Leaf implements ScorerAware {
+        /**
+         * Collect this doc, returning {@code true} if it is competitive.
+         */
+        public final boolean hit(int doc, long bucket) throws IOException {
```

Comment:
Is the "hit" naming following a search/lucene convention?  If not, I wonder if we should call it `collect()` to mirror the agg conventions?  It's obviously not quite the same since it returns a bool, but it does "collect" at the same time too.  Dunno, no strong opinion.

Reply:
It only `collect`s if the hit is competitive so I didn't like `collect`. I'll think on the name!

Reply:
aha right, makes sense.  `collectIfCompetitive()`?  fwiw, other aggs discard docs in `collect()` if they aren't "competitive" so not too unusual, e.g. min/max obviously only keep min or max, bucketing aggs can discard docs if they don't match include/exclude, etc etc.  

But they also don't have a return value, so it is probably good to call this something different just so it's clear they aren't exactly the same

I dunno, no strong feelings here :)

Reply:
I'm going to just go with `collectIfCompetitive` for now. I'll have to add priority queue stuff in a follow up and then the method's name may scream at me to change. But for now `collect` seems fine.

Reply:
Works for me!